src/common/walFilter/versions/recordProcessGPDB6.c

```diff
@@ -0,0 +1,620 @@
+{
+    BlockIdData ip_blkid;
+    OffsetNumber ip_posid;
+}
+
+#ifdef __arm__
+__attribute__((packed))         /* Appropriate whack upside the head for ARM */
+#endif
+ItemPointerData;
+
+typedef struct xl_heaptid
+{
+    RelFileNode node;
+    ItemPointerData tid;        /* changed tuple id */
+} xl_heaptid;
+
+typedef struct xl_heap_new_cid
+{
+    /*
+     * store toplevel xid so we don't have to merge cids from different
+     * transactions
+     */
+    TransactionId top_xid;
+    CommandId cmin;
+    CommandId cmax;
+
+    /*
+     * don't really need the combocid since we have the actual values right in
+     * this struct, but the padding makes it free and its useful for
+     * debugging.
+     */
+    CommandId combocid;
+
+    /*
+     * Store the relfilenode/ctid pair to facilitate lookups.
+     */
+    xl_heaptid target;
+} xl_heap_new_cid;
+
+static
+bool
+getXlog(const XLogRecord *record, RelFileNode **node)
+{
+    const uint8_t info = (uint8_t) (record->xl_info & ~XLR_INFO_MASK);
+    switch (info)
+    {
+        case XLOG_CHECKPOINT_SHUTDOWN:
+        case XLOG_CHECKPOINT_ONLINE:
+        case XLOG_NOOP:
+        case XLOG_NEXTOID:
+        case XLOG_NEXTRELFILENODE:
+        case XLOG_RESTORE_POINT:
+        case XLOG_BACKUP_END:
+        case XLOG_PARAMETER_CHANGE:
+        case XLOG_FPW_CHANGE:
+        case XLOG_END_OF_RECOVERY:
+        case XLOG_OVERWRITE_CONTRECORD:
+        case XLOG_SWITCH:
+//          ignore
+            break;
+
+        case XLOG_FPI:
+        {
+            *node = (RelFileNode *) XLogRecGetData(record);
+            return true;
+        }
+
+        default:
+            THROW_FMT(FormatError, "XLOG UNKNOWN: %d", info);
+    }
+    return false;
+}
+
+static
+bool
+getStorage(const XLogRecord *record, RelFileNode **node)
+{
+    const uint8_t info = (uint8_t) (record->xl_info & ~XLR_INFO_MASK);
+    switch (info)
+    {
+        case XLOG_SMGR_CREATE:
+        {
+            *node = (RelFileNode *) XLogRecGetData(record);
+            return true;
+        }
+
+        case XLOG_SMGR_TRUNCATE:
+        {
+            xl_smgr_truncate *xlrec = (xl_smgr_truncate *) XLogRecGetData(record);
+
+            *node = &xlrec->rnode;
+            return true;
+        }
+
+        default:
+            THROW_FMT(FormatError, "Storage UNKNOWN: %d", info);
+    }
+}
+
+static
+bool
+getHeap2(const XLogRecord *record, RelFileNode **node)
+{
+    uint8_t info = (uint8_t) (record->xl_info & ~XLR_INFO_MASK);
+    info &= XLOG_HEAP_OPMASK;
+
+    if (info == XLOG_HEAP2_NEW_CID)
+    {
+        xl_heap_new_cid *xlrec = (xl_heap_new_cid *) XLogRecGetData(record);
+        *node = &xlrec->target.node;
+        return true;
+    }
+
+    if (info == XLOG_HEAP2_REWRITE)
+    {
+        return false;
+    }
+
+    // XLOG_HEAP2_CLEAN
+    // XLOG_HEAP2_FREEZE_PAGE
+    // XLOG_HEAP2_CLEANUP_INFO
+    // XLOG_HEAP2_VISIBLE
+    // XLOG_HEAP2_MULTI_INSERT
+    // XLOG_HEAP2_LOCK_UPDATED
```

Comment:
why not case with default error?

Reply:
Because all possible values are used. `XLOG_HEAP_OPMASK` has three non-zero bits, which gives 8 possible values, all of which are used.

Reply:
ok, how about getHeap?

Reply:
The same thing.

Reply:
> 0x030 is free, was XLOG_HEAP_MOVE